dot/parachain/util/backing_implicit_view.go

```diff
@@ -0,0 +1,565 @@
+		return err
+	}
+
+	// Retain at least `minimumRetainLength` blocks in storage.
+	// This helps to avoid Chain API calls when activating leaves in the same chain.
+	retainMinimum := min(
+		fetched.minimumAncestorNumber,
+		max(fetched.leafNumber-minimumRetainLength, 0),
+	)
+	// store the leaf in the active leaves map
+	view.leaves[leafHash] = activeLeafPruningInfo{retainMinimum: retainMinimum}
+	return nil
+}
+
+// Deactivate a leaf in the view. This prunes any outdated implicit ancestors as well.
+// Returns hashes of blocks pruned from storage.
+func (view *BackingImplicitView) DeactivateLeaf(leafHash common.Hash) []common.Hash {
+	// Check if the leaf exists in the active leaves map
+	if _, exists := view.leaves[leafHash]; !exists {
+		return nil
+	}
+
+	// Remove the leaf from the active leaves map
+	delete(view.leaves, leafHash)
+
+	// Determine the minimum retain block number across all remaining leaves
+	minimumBlockNumber := ^parachaintypes.BlockNumber(0) // Max value for BlockNumber
+	for _, leafInfo := range view.leaves {
+		if leafInfo.retainMinimum < minimumBlockNumber {
+			minimumBlockNumber = leafInfo.retainMinimum
+		}
+	}
+
+	// Prune everything before the minimum out of all leaves,
+	// pruning absolutely everything if there are no leaves (empty view)
+
+	// Collect and prune blocks with blockNumber < minimum
+	var leafToRemove []common.Hash
+	for hash, info := range view.blockInfoStorage {
+		if info.blockNumber < minimumBlockNumber {
+			leafToRemove = append(leafToRemove, hash)
+		}
+	}
+
+	for _, hash := range leafToRemove {
+		delete(view.blockInfoStorage, hash)
+	}
+
+	return leafToRemove
+}
+
+// KnownAllowedRelayParentsUnder returns known, allowed relay-parents that are valid for parachain candidates
+// which could be backed in a child of a given block for a given para ID.
+//
+// This is expressed as a contiguous slice of relay-chain block hashes which may
+// include the provided block hash itself.
+//
+// If `paraID` is nil, this returns all valid relay-parents across all paras for the leaf.
+func (view *BackingImplicitView) KnownAllowedRelayParentsUnder(
+	blockHash common.Hash, paraID *parachaintypes.ParaID) []common.Hash {
+	blockInfo, exists := view.blockInfoStorage[blockHash]
+	if !exists {
+		return nil
+	}
+
+	return blockInfo.allowedRelayParents.allowedFor(paraID, blockInfo.blockNumber)
+}
+
+func (view *BackingImplicitView) fetchFreshLeafAndInsertAncestry(
+	leafHash common.Hash, subsystemToOverseer chan<- any) (*fetchSummary, error) {
+	// Get the block header directly from block state
+	blockHeader, err := view.blockState.GetHeader(leafHash)
+	if err != nil {
+		return nil, fmt.Errorf("getting block header for leaf %s: %w", leafHash, err)
+	}
+
+	minRelayParents, err := view.findMinRelayParents(leafHash, blockHeader, subsystemToOverseer)
+	if err != nil {
+		return nil, fmt.Errorf("getting min relay parents for leaf %s: %w", leafHash, err)
+	}
+
+	minAncestorNumber := ^parachaintypes.BlockNumber(0)
```

Comment:
two questions here:
1. `^` operator is flipping the bits so it should get the max value of uint32(the actual type of BlockNumber), why not just use `math.MaxUint32`?
2. it returns the max value but the var is called `min....`, it seems to be conflict?

Reply:
Yeah we can use math.MaxUint32.

Also it is a widely used approach to initialise the variable to its maximum possible value if we want to find & store the minimum.